File: typescript/src/create-party.ts
import { createFrom } from 'stedy/bytes'
import { KeyPair, Party } from '../types'
import createCalculateSafetyNumber from './create-calculate-safety-number'
import createHandshake from './create-handshake'
import generateEphemeralKeyPair from './generate-ephemeral-key-pair'

const createParty = async (
  isInitiator: boolean,
  identityKeyPair: KeyPair
): Promise<Party> => {
  const ephemeralKeyPair = await generateEphemeralKeyPair()
  const calculateSafetyNumber = createCalculateSafetyNumber(
    identityKeyPair.publicKey
  )
  const handshake = createHandshake(
    isInitiator,
    identityKeyPair,
    ephemeralKeyPair
  )
  return {
    calculateSafetyNumber,
    ephemeralKey: createFrom(ephemeralKeyPair.publicKey),
    handshake,
    identityKey: createFrom(identityKeyPair.publicKey)
  }
}

export default createParty

File: cplusplus/src/create_party.cpp
#include "autograph/create_party.h"

Party create_party(bool is_initiator, const KeyPair &identity_key_pair) {
  KeyPair ephemeral_key_pair = generate_ephemeral_key_pair();
  CalculateSafetyNumberFunction calculate_safety_number =
      create_calculate_safety_number(identity_key_pair.public_key);
  HandshakeFunction handshake =
      create_handshake(is_initiator, identity_key_pair, ephemeral_key_pair);
  Party party = {
      calculate_safety_number,
      ephemeral_key_pair.public_key,
      handshake,
      identity_key_pair.public_key,
  };
  return std::move(party);
}
